[#1369] Implement basic vertex deduplication.

p-f opened the pull request:
This selects certain vertices by label and deduplicates them based on equal properties.
Edges will be updated.

fixes #1369

timo95 commented on gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/VertexDeduplication.java:
```diff
@@ -0,0 +1,113 @@
+      .reduceGroup(new MarkDuplicatesInGroup<>())
+      .name("Mark duplicate vertices");
+    DataSet<Tuple2<GradoopId, GradoopId>> vertexToDedupVertex = annotatedVertices
+      .flatMap(new CreateMappingFromMarkedDuplicates<>())
```
I think you can save several flink operators by creating the mapping in the `reduceGroup` step and writing it as map property on each original element. Then you can remove the duplicates directly in the `reduceGroup` as well. The data flow should be similar, but you reduce the overhead.

p-f replied:
I think this would destroy scalablity because in the worst case all vertices are in one group, leading to one element storing all other ids.
Storing the mapping on vertices should be good for many small groups, but I don't think that we should assume this

timo95 commented on gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/VertexDeduplication.java:
```diff
@@ -0,0 +1,113 @@
+      .flatMap(new CreateMappingFromMarkedDuplicates<>())
+      .name("Create mapping to deduplicated vertices");
+    DataSet<V> deduplicatedVertices = annotatedVertices.filter(new RemoveMarkedDuplicates<>());
+    DataSet<V> otherVertices = graph.getVertices().filter(new Not<>(new ByLabel<>(label)));
```
I think `.negate()` would be better.

timo95 commented on gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/VertexDeduplication.java:
```diff
@@ -0,0 +1,113 @@
+      .leftOuterJoin(vertexToDedupVertex)
+      .where(new TargetId<>())
+      .equalTo(0)
+      .with(new UpdateEdgeTarget<>())
```
Annotating the update functions might enable flink to do both updates in parallel:
`@FunctionAnnotation.NonForwardedFields("targetId")`
Does this actually work with the interfaces? After all there is no "targetId" field...

p-f replied:
NonForwardedFields is not available for join functions, I added other annotations instead.

timo95 commented on gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/VertexDeduplication.java:
```diff
@@ -0,0 +1,113 @@
+    DataSet<Tuple2<GradoopId, GradoopId>> vertexToDedupVertex = annotatedVertices
+      .flatMap(new CreateMappingFromMarkedDuplicates<>())
+      .name("Create mapping to deduplicated vertices");
+    DataSet<V> deduplicatedVertices = annotatedVertices.filter(new RemoveMarkedDuplicates<>());
```
`new ByProperty(MarkDuplicatesInGroup.PROPERTY_KEY).negate()`

timo95 commented on gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/functions/RemoveMarkedDuplicates.java:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @param <E> The element type.
+ */
+public class RemoveMarkedDuplicates<E extends Element> implements CombinableFilter<E> {
```
not necessary (see above)

timo95 commented on gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/functions/UpdateEdgeSource.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @param <E> The edge type.
+ */
+public class UpdateEdgeSource<E extends Edge> implements JoinFunction<E, Tuple2<GradoopId, GradoopId>, E> {
```
Almost the same code as `EdgeSourceUpdateJoin`. Consider removing one and keeping the other with the null check and in a fitting location.

p-f replied:
Those functions are similar, but the other functions are part of the cloning operator. We should consider unifying the join functions, but since this would affect another operator, I think I'll leave this for now.

ChrizZz110 replied:
I agree to @timo95 . Please move EdgeSourceUpdateJoin and EdgeTargetUpdateJoin to package `org.gradoop.flink.model.impl.functions.epgm` and use it here instead.

timo95 commented on gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/functions/UpdateEdgeTarget.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @param <E> The edge type.
+ */
+public class UpdateEdgeTarget<E extends Edge> implements JoinFunction<E, Tuple2<GradoopId, GradoopId>, E> {
```
Almost the same code as `EdgeTargetUpdateJoin`. See above.

ChrizZz110 replied:
Same here

timo95 commented on gradoop-data-integration/src/test/java/org/gradoop/dataintegration/transformation/VertexDeduplicationTest.java:
```diff
@@ -0,0 +1,80 @@
+    FlinkAsciiGraphLoader loader = getLoaderFromString("input[" +
+      "(:a {key: 1L})-[:e {att: 1L}]->(:a {key: 1L})-[:e {att: 2L}]->(:b {key: 1L})" +
+      "-[:e {att: 3L}]->(:a {key: 2L})(:a {key: 2L})" +
+      "]" +
```
This example is too simple. You can test it in one graph, but at least add multiple keys and different not duplicated vertices (e.g. with one property equal and one missing or the same key but different values/types, etc).

p-f replied:
I changed it to a more complex example. Cases for all missing keys etc. are tested with the key extractor function

timo95 commented on gradoop-data-integration/src/test/java/org/gradoop/dataintegration/transformation/VertexDeduplicationTest.java:
```diff
@@ -0,0 +1,80 @@
+   * @throws Exception when the execution in Flink fails.
+   */
+  @Test
+  public void testWithUnknownType() throws Exception {
```
`testWithUnknownLabel`
For me type is int, float etc.

ChrizZz110 commented:
Please remove naming of transformations.

timo95 commented on gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/functions/UpdateEdgeTarget.java:
```diff
@@ -30,6 +31,8 @@
  *
  * @param <E> The edge type.
  */
+@FunctionAnnotation.ForwardedFieldsFirst("id;label;properties;sourceId;graphIds")
```
`NonForwardedFieldsFirst` would be shorter and clearer.

timo95 commented on gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/functions/UpdateEdgeSource.java:
```diff
@@ -30,6 +31,8 @@
  *
  * @param <E> The edge type.
  */
+@FunctionAnnotation.ForwardedFieldsFirst("id;label;properties;targetId;graphIds")
```
`NonForwardedFieldsFirst` would be shorter and clearer.

p-f commented:
Done, I've moved the functions to the `epgm.functions` package and moved the tests accordingly. I had to make some minor changes, which do not affect the cloning operator though.

ChrizZz110 commented:
Please also add a small Wiki-Entry when this is merged.
